// hapi/lib/ext.js
'use strict';

// Load modules

var Hoek = require('hoek');
var Topo = require('topo');


// Declare internals

var internals = {};


exports = module.exports = internals.Ext = function (type, server) {

    this._topo = new Topo();
    this._server = server;
    this._routes = [];

    this.type = type;
    this.nodes = null;
};


internals.Ext.prototype.add = function (event) {

    var methods = [].concat(event.method);
    var options = event.options;

    for (let i = 0; i < methods.length; ++i) {
        var settings = {
            before: options.before,
            after: options.after,
            group: event.plugin.realm.plugin,
            sort: this._server._extensionsSeq++
        };

        var node = {
            func: methods[i],                 // Connection: function (request, next), Server: function (server, next)
            bind: options.bind,
            plugin: event.plugin
        };

        this._topo.add(node, settings);
    }

    this.nodes = this._topo.nodes;

    // Notify routes

    for (let i = 0; i < this._routes.length; ++i) {
        this._routes[i].rebuild(event);
    }
};


internals.Ext.prototype.merge = function (others) {

    var merge = [];
    for (let i = 0; i < others.length; ++i) {
        merge.push(others[i]._topo);
    }

    this._topo.merge(merge);
    this.nodes = (this._topo.nodes.length ? this._topo.nodes : null);
};


internals.Ext.prototype.subscribe = function (route) {

    this._routes.push(route);
};


internals.Ext.combine = function (route, type) {

    var ext = new internals.Ext(type, route.server);

    var events = route.settings.ext[type];
    if (events) {
        for (let i = 0; i < events.length; ++i) {
            var event = Hoek.shallow(events[i]);
            Hoek.assert(!event.options.sandbox, 'Cannot specify sandbox option for route extension');
            event.plugin = route.plugin;
            ext.add(event);
        }
    }

    var connection = route.connection._extensions[type];
    var realm = route.plugin.realm._extensions[type];

    ext.merge([connection, realm]);

    connection.subscribe(route);
    realm.subscribe(route);

    return ext;
};
